### PR TITLE
Allow external frontend development

### PR DESCRIPTION
**Pull request by GabrielL**
This patchset aims to simplify a bit the Language loading facilities, and allow external language providers.

## Introducing SPI for Language frontends

It introduces a new interface, [`de.jplag.LanguageProvider`](https://github.com/GabrielL/JPlag/blob/cc51c80d36fdf950c7b24a706796695e140a821b/jplag.frontend-utils/src/main/java/de/jplag/LanguageProvider.java) (in frontend-utils) that is implemented by every language frontend.

For example, for [java](https://github.com/GabrielL/JPlag/blob/cc51c80d36fdf950c7b24a706796695e140a821b/jplag.frontend.java/src/main/java/de/jplag/java/JavaLanguageProvider.java) :

```java
@MetaInfServices
public class JavaLanguageProvider implements LanguageProvider {
    public static final String NAME = de.jplag.java.Language.NAME;

    @Override
    public Language create(final ErrorConsumer errorCollector) {
        return new de.jplag.java.Language(errorCollector);
    }

    @Override
    public String getDisplayName() {
        return NAME;
    }
}
```

With this, we can use the [Java Service Loader facilities](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html) to discover frontend in the class path, instead of having to add them inside the main repository.

All the loading, and usage of the Service Loader is located inside [`de.jplag.Languages`](https://github.com/GabrielL/JPlag/blob/cc51c80d36fdf950c7b24a706796695e140a821b/jplag/src/main/java/de/jplag/Languages.java)

In order to be discoverable, this also adds a `META-INF/services/de.jplag.LanguageProvider` file inside the jar (example for java: https://github.com/GabrielL/JPlag/blob/cc51c80d36fdf950c7b24a706796695e140a821b/jplag.frontend.java/src/main/resources/META-INF/services/de.jplag.LanguageProvider)

## Refactoring and other changes

In order to make these changes, some refactoring were done:

* simplification of options/language usage in report, both are now directly passed to the Report constructor.
* removal of `language` in `JPlagOption`, options are now nearly read only.
* removal of `JPlagOptions` in `JPlagResult`
* removal of `LanguageOption`, its work is covered by `Languages` and `LanguageProvider` SPI.
* small refactoring in `JPlag` class, mostly making all fields final where possible

## Build Changes

In order to have working service providers, jars must have a `META-INF/services` directory with all the services exposed.

For the assembly jar with all dependencies, there is some changes to be able to merge all the `META-INF/services` files inside all the frontend jars.

### CODE REVIEW

**Comment by dfuchss** [comment]
@GabrielL seems to be a nice enhancement.

One suggestion: 
If you are using [ServiceLoader](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html) (and you are using it :) ), you might also use [MetaInfServices](https://mvnrepository.com/artifact/org.kohsuke.metainf-services/metainf-services/1.8) (see also: [Docs](http://metainf-services.kohsuke.org/)). Actually it generates the `META-INF/services/de.jplag.LanguageProvider` automatically :)

**Comment by GabrielL** [comment]
> @GabrielL seems to be a nice enhancement.
> 
> One suggestion: If you are using [ServiceLoader](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html) (and you are using it :) ), you might also use [MetaInfServices](https://mvnrepository.com/artifact/org.kohsuke.metainf-services/metainf-services/1.8) (see also: [Docs](http://metainf-services.kohsuke.org/)). Actually, it generates the `META-INF/services/de.jplag.LanguageProvider` automatically :)

I didn't know this one, this is nicer like this, updated, thanks!

File: jplag/src/main/assembly/jar-with-deps-merge-services.xml
Comment (tsaglam) [inline jplag/src/main/assembly/jar-with-deps-merge-services.xml:6]: Jar name should still be `jar-with-dependencies` so adapt accordingly if required.

File: jplag/src/main/java/de/jplag/JPlag.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/JPlag.java:110]: This became so short, there might not be a need for a dedicated method.

File: jplag/src/main/java/de/jplag/Languages.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/Languages.java:37]: Please avoid nested classes!

File: README.md
Comment (tsaglam) [reply README.md:66]: Actually, I am now wondering how you would use the JPlag API during 3rd party integration.
Imagine an application that depends on JPlag and only ever runs it with a specific language.
Right now, how would you specify here to run JPlag with C++?

**Review by tsaglam — CHANGES_REQUESTED** [review]
There are some issues that need to be adressed.

**Review by dfuchss — CHANGES_REQUESTED** [review]
Instead of using the xml to create the assembly, you may have a look at https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ServicesResourceTransformer

File: jplag/src/main/assembly/jar-with-deps-merge-services.xml
Comment (GabrielL) [reply jplag/src/main/assembly/jar-with-deps-merge-services.xml:6]: Since there is already a descriptor `jar-with-dependencies` I can't reuse the name, hence the changed name.

Initially I wanted to avoid adding a new dependency into the build, just to do that, but I'll look into shade to come up with something else.

File: jplag/src/main/java/de/jplag/JPlag.java
Comment (GabrielL) [reply jplag/src/main/java/de/jplag/JPlag.java:110]: This is the factory, I know this is short, but the goal here is to document the line, and hide construction details.

I will try to find another way to do that, like integrate it directly into `Languages`

File: README.md
Comment (tsaglam) [reply README.md:66]: I am not a big fact that the users of the API need to enter random strings. This was one of the reasons for the language options enum. Maybe we can restore the enum (without the classpath field of course), just to give the users some options here. Users can then still specify the string or they can just pick one of the enumerals in the language option enum. What do you think?

Comment (dfuchss) [reply README.md:66]: you may also define constants in the languages itself that can be used as keys. you could also replace the constant DefaultLanguage with JavaLanguageProvider.NAME

File: .github/workflows/maven.yml
Comment (GabrielL) [reply .github/workflows/maven.yml]: fixed using shade instead of the assembly plugin.

As a side effect, it is no longer necessary to call specifically `assembly:single` in order to have the uber jar, this is already included in the `package` phase

File: jplag/pom.xml
Comment (GabrielL) [inline jplag/pom.xml:60]: The change is here.

It seems that the file was using CRLF lines instead of LF only

Comment (dfuchss) [inline jplag/pom.xml:60]: See above. I would not create that by default.

File: jplag/src/main/java/de/jplag/Languages.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/Languages.java:21]: maybe throw an exception/ log an error if a name is duplicated.

File: jplag/src/main/java/de/jplag/reporting/Report.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/Report.java:75]: I think you should delete private getters. They are not needed.

File: pom.xml
Comment (dfuchss) [inline pom.xml:99]: I would also remove the version of the standard maven plugins.
The current versions are inherited by the maven super pom. if there is no reason to fix them to a certain version.

File: pom.xml
Comment (dfuchss) [inline pom.xml:112]: see above

File: pom.xml
Comment (dfuchss) [inline pom.xml:125]: see above

File: pom.xml
Comment (dfuchss) [inline pom.xml:191]: see above

**Review by dfuchss — CHANGES_REQUESTED** [review]
Nice work! :)

I've found some points for discussion and some points that maybe could be improved :)

File: pom.xml
Comment (GabrielL) [reply pom.xml:99]: maven is complaining about it during the build, so it seems that it needs/wants it. I don't know if it is really useful or not

File: jplag/src/main/java/de/jplag/reporting/Report.java
Comment (GabrielL) [reply jplag/src/main/java/de/jplag/reporting/Report.java:75]: Yes, they are not needed. Their main usage here was during refactoring, I will inline them

File: .github/workflows/maven.yml
Comment (GabrielL) [reply .github/workflows/maven.yml:28]: Why split the jar generation in two phases ? Here there is a known command `mvn package` that will build the needed jars, and no need to search the documentation for which plugin/command is used in this specific project.

File: jplag.frontend.chars/src/main/java/de/jplag/chars/Language.java
Comment (GabrielL) [reply jplag.frontend.chars/src/main/java/de/jplag/chars/Language.java:13]: There is no documentation in all the Language class

File: jplag/src/main/java/de/jplag/Languages.java
Comment (GabrielL) [reply jplag/src/main/java/de/jplag/Languages.java:21]: Oh yes, good catch, now it is possible to have duplicated names

File: jplag/src/main/java/de/jplag/Languages.java
Comment (GabrielL) [inline jplag/src/main/java/de/jplag/Languages.java:9]: documentation is missing

File: .github/workflows/maven.yml
Comment (dfuchss) [reply .github/workflows/maven.yml:28]: Alternatively you could move that to a profile that you can control the build process. (If I invoke package I assume maven to build the library artifacts nothing more .. but that's mainly  personal expectation .. I will / we should have a look at the docs how they define the package phase )

I personally like the idea that 'package' build the necessary stuff. Here that would only be the libraries that are going to be deployed to maven central. For the complete jar. You could define the Profile "XYZ" that can be used or not by the users of JPlag.


That's something the maintainers should decide in the end :)

Comment (tsaglam) [reply .github/workflows/maven.yml:28]: I second what @dfuchss said. With `package` we build the jars for maven central, with `assembly:single` we build the fat jar. In general, details regarding the process should be left untouched by PRs if possible. Meaning how to build JPlag, how the JAR is named, and even how to use JPlag itself. If something _needs_ to be changed, it should be declared explicitly in the PR description.

File: jplag.frontend.chars/src/main/java/de/jplag/chars/Language.java
Comment (dfuchss) [reply jplag.frontend.chars/src/main/java/de/jplag/chars/Language.java:13]: Yes, but that are methods that are inherited from the `de.jplag.Language`.
You introduce a new public constant. Therefore, (in my opinion) its purpose should be documented :)

**Comment by tsaglam** [comment]
There are conflicts that need to be resolved before we can proceed. Also, please mark all suggestions as resolved if you addressed them.

**Comment by tsaglam** [comment]
Closed due to lack of activity, happy to re-open anytime.

**Pull request closed without merging**
